dla_cnn/desi/preprocess.py

```diff
@@ -92,3 +92,68 @@ def label_sightline(sightline, kernel, REST_RANGE, pos_sample_kernel_percent=0.3
     # classification is 1 / 0 / -1 for DLA/nonDLA/border
     # offsets_array is offset
     return classification, offsets_array, column_density
+
+def rebin(sightline, v = 20000):
+    """
+    Resample and rebin the input Sightline object's data to a constant dlambda/lambda dispersion.
+
+    Parameters
+    ----------
+    sightline: dla_cnn.data_model.Sightline
+    v: float
+    
+    Returns
+    -------
+    sightline: dla_cnn.data_model.Sightline
+    """
+    c = 2.9979246e8
+    dlambda = np.log(1+v/c)
+    wavelength = np.exp(np.log(10)*sightline.loglam)
+    central_wavelength = [item.central_wavelength for item in sightline.dlas]
+    max_wavelength = wavelength[-1]/min(central_wavelength)
+    min_wavelength = wavelength[0]/max(central_wavelength)
+    pixels_number = int(np.round(np.log(max_wavelength/min_wavelength)/dlambda))+1
+    new_wavelength = wavelength[0]*np.exp(dlambda*np.arange(pixels_number))
+    
+    npix = len(wavelength)
+    wvh = (wavelength + np.roll(wavelength, -1)) / 2.
+    wvh[npix - 1] = wavelength[npix - 1] + \
+                    (wavelength[npix - 1] - wavelength[npix - 2]) / 2.
+    dwv = wvh - np.roll(wvh, 1)
+    dwv[0] = 2 * (wvh[0] - wavelength[0])
+    med_dwv = np.median(dwv)
+    
+    cumsum = np.cumsum(sightline.flux * dwv)
+    cumvar = np.cumsum(sightline.error * dwv, dtype=np.float64)
+    
+    fcum = interp1d(wvh, cumsum,bounds_error=False)
+    fvar = interp1d(wvh, cumvar,bounds_error=False)
+    
+    nnew = len(new_wavelength)
+    nwvh = (new_wavelength + np.roll(new_wavelength, -1)) / 2.
+    nwvh[nnew - 1] = new_wavelength[nnew - 1] + \
+                     (new_wavelength[nnew - 1] - new_wavelength[nnew - 2]) / 2.
+    
+    bwv = np.zeros(nnew + 1)
+    bwv[0] = new_wavelength[0] - (new_wavelength[1] - new_wavelength[0]) / 2.
+    bwv[1:] = nwvh
+    
+    newcum = fcum(bwv)
+    newvar = fvar(bwv)
+    
+    new_fx = (np.roll(newcum, -1) - newcum)[:-1]
+    new_var = (np.roll(newvar, -1) - newvar)[:-1]
+    
+    # Normalize (preserve counts and flambda)
+    new_dwv = bwv - np.roll(bwv, 1)
+    new_fx = new_fx / new_dwv[1:]
+    # Preserve S/N (crudely)
+    med_newdwv = np.median(new_dwv)
+    new_var = new_var / (med_newdwv/med_dwv) / new_dwv[1:]
+    
+    test = (~np.isnan(new_fx))|(~np.isnan(new_var))
+    sightline.loglam = np.log10(new_wavelength[test])
```

Comment:
remove `test` from these lines.

Reply:
In fact, I think maybe we should remain the `test` now, or we have to deal with those out of range points. Maybe there will be better solutions, I will have a try. Hope for your advice.

Reply:
Well, can you first have the code check that the `nan` are at the very edges of
the spectrum.  If they are, then I am ok with keeping those last lines of code as is.

Reply:
Ok, I will do it. Thanks for your advice.

Reply:
I have tested the `nan` are at the very edges of the spectrum, and if the spectrum have some `nan` value in the main part, the function will raise error. Hope for your advice.